lib/luanetfilter.c

```diff
@@ -35,15 +36,115 @@ typedef struct luanetfilter_xtable_s {
 	lunatik_object_t *runtime;
 } luanetfilter_xtable_t;
 
-static int luanetfilter_xtable(lua_State *L);
+static struct xt_match *match = NULL;
+static struct xt_target *target = NULL;
+
+static void luanetfilter_release(void *private);
+static int luanetfilter_xtable_match(lua_State *L);
+static int luanetfilter_xtable_target(lua_State *L);
+
+#define LUANETFILTER_XTABLE_REG(mode)						\
+static int luanetfilter_xtable_##mode(lua_State *L) {				\
+	lunatik_object_t *object;						\
+	luanetfilter_xtable_t *luaxtable;					\
+										\
+	object = lunatik_newobject(L, &luanetfilter_class, sizeof(luanetfilter_xtable_t));	\
+	luaxtable = (luanetfilter_xtable_t *)object->private;			\
+										\
+	luaxtable->runtime = lunatik_toruntime(L);				\
+	lunatik_getobject(luaxtable->runtime);					\
+										\
+	luaL_checktype(L, 1, LUA_TTABLE);					\
+										\
+	luanetfilter_xtable_opt(L, 1, "name", LUA_TSTRING, true);		\
+	const char *name = lua_tostring(L, -1);					\
+										\
+	luanetfilter_xtable_opt(L, 1, "revision", LUA_TNUMBER, true);		\
+	uint8_t revision = (uint8_t)lua_tointeger(L, -1);			\
+										\
+	luanetfilter_xtable_opt(L, 1, "family", LUA_TNUMBER, true);		\
+	unsigned short family = (unsigned short)lua_tointeger(L, -1);		\
+										\
+	unsigned proto = luanetfilter_xtable_opt(L, 1, "proto", LUA_TNUMBER, false) ? \
+			 (unsigned short)lua_tointeger(L, -1) : 0;		\
+										\
+	luanetfilter_xtable_opt(L, 1, "checkentry", LUA_TFUNCTION, false);	\
+										\
+	luanetfilter_xtable_opt(L, 1, "destroy", LUA_TFUNCTION, false);		\
+	luanetfilter_xtable_opt(L, 1, #mode, LUA_TFUNCTION, false);		\
+										\
+	mode = lunatik_checkalloc(L, sizeof(struct xt_##mode));			\
+	if (!mode) {								\
+		luaL_error(L, "failed to allocate memory for" #mode);		\
+	}									\
+										\
+	memcpy((void *)mode->name, name, XT_EXTENSION_MAXNAMELEN);		\
+	mode->revision = revision;						\
+	mode->family = family;							\
+	mode->proto = proto;							\
+	mode->mode = luanetfilter_##mode;					\
+	mode->checkentry = luanetfilter_##mode##_check;				\
+	mode->destroy = luanetfilter_##mode##_destroy;				\
+	mode->me = THIS_MODULE;							\
+										\
+	if (xt_register_##mode(mode) != 0) {					\
+		lunatik_free(mode);						\
+		luaL_error(L, "unable to register" #mode);			\
+	}									\
+										\
+	lunatik_registerobject(L, 1, object);					\
+	return 1;								\
+}
 
-static void luanetfilter_release(void *private)
+/* Verifies the type of "opt" in the table at "idx" and returns true if the value can be extracted.
+ * If the value is not required, it returns false.
+ * If the value is required and the type is invalid, it raises an error.
+ * */
+static bool luanetfilter_xtable_opt(lua_State *L, int idx, const char *opt, int type, bool req)
 {
-	luanetfilter_xtable_t *luaxtable = (luanetfilter_xtable_t *)private;
-	lunatik_putobject(luaxtable->runtime);
+	lua_getfield(L, idx, opt);
+	int type_ = lua_type(L, -1);
```

Comment:
it looks like you are kinda reimplementing `luaL_check*` APIs.. please take a look at the reference manual and verify if none suits to you..

Reply:
I needed this because some parameters are optional and some are mandatory, so for optional parameters we can use a default value.

So, if the check is successful -> returns true -> take the value from the stack
if the check is unsuccessful -> it is required -> raises luaL_error
if the check is unsuccessful -> it is _not_ required -> returns false -> pick a default value

Reply:
can't you use [luaL_opt*](https://www.lua.org/manual/5.4/manual.html#luaL_opt)?

Reply:
we might at least use their implementation as a base to implement this function.. replicating the same idea but for table fields..

Reply:
For datatypes other than functions, I have proceeded as follows. Is this fine ?? : 
```c
	lua_getfield(L, 1, "name");				   		\
	const char *name = luaL_checkstring(L, -1);				\
										\
	lua_getfield(L, 1, "revision");				   		\
	uint8_t revision = (uint8_t)luaL_checkinteger(L, -1);			\
										\
	lua_getfield(L, 1, "family");				   		\
	unsigned short family = (unsigned short)luaL_checkinteger(L, -1);	\
                                                                                \
	/*optional parameter*/							\
	lua_getfield(L, 1, "proto");				   		\
	unsigned short proto = (unsigned short)luaL_optinteger(L, -1, 0);	\
```

For functions I use the following (the functions can be nil and no default value is needed, but they should be valid) :
```c
#define luanetfilter_xtable_func(L, idx, name)					\
do {  								          	\
	lua_getfield(L, idx, name);				   	  	\
	if(!lua_isnoneornil(L, -1)) luaL_checktype(L, -1, LUA_TFUNCTION); 	\
} while(0)

...

luanetfilter_xtable_func(L, 1, "checkentry");				\
luanetfilter_xtable_func(L, 1, "destroy");				\
```

Reply:
> For datatypes other than functions, I have proceeded as follows. Is this fine ?? :
> ...

the only issue I see with this approach is the error messages will be a little misleading.. it will refer to a unknown argument.. I think we can leave this as is and add a TODO comment to create a `lunatik_checkfield()` API to check the type of a table field and raise a proper error message.. something like [this](https://github.com/luainkernel/lunatik/blob/master/lib/luadevice.c#L289).. 

> For functions I use the following (the functions can be nil and no default value is needed, but they should be valid) :
> ...

why does the callback can be nil? what's the use case for that?

Reply:
If the user doesn't want to have a checkentry ir destroy function, it can be nil.

Reply:
I would simplify by now and force the user to define at least a `nop` callback as we do for the device library.. if we really have practical use case, we can always do this in a higher layer implementing it in Lua itself.. let's keep the binding as simpler as possible ;-)

Reply:
Ok, understood